README.rst
==============
Pyramid_health
==============

Simple healthcheck endpoint for Pyramid, with maintenance mode and application
checks.

* PyPI: https://pypi.python.org/pypi/pyramid_health
* Github: https://github.com/ludia/pyramid_health
* |droneio|

.. |droneio| image::
   https://drone.io/github.com/ludia/pyramid_health/status.png
   :target: https://drone.io/github.com/ludia/pyramid_health
   :alt: Tests on drone.io


Installation
============

Install using setuptools, e.g. (within a virtualenv)::

  $ pip install pyramid_health


Setup
=====

Once ``pyramid_health`` is installed, you must use the ``config.include``
mechanism to include it into your Pyramid project's configuration.  In your
Pyramid project's ``__init__.py``:

.. code-block:: python

   config = Configurator(.....)
   config.include('pyramid_health')

Alternately you can use the ``pyramid.includes`` configuration value in your
``.ini`` file:

.. code-block:: ini

   [app:myapp]
   pyramid.includes = pyramid_health


Usage
=====

Pyramid_health configuration (values are defaults):

.. code-block:: ini

   [app:myapp]
   healthcheck.url = /health

   healthcheck.disablefile = /tmp/maintenance  # touch this file to activate

   healthcheck.maintenance_code = 299  # Code to return in maintenance mode

   healthcheck.failure_code = 503  # Code to return when one or more checks fail


Operation
=========

When your application is healthy, pyramid_health endpoint returns ``200 OK``.
When you enable the maintenance mode, the endpoint returns ``299 MAINTENANCE``
and logs ``Health response: MAINTENANCE``.
If the request to the healthcheck endpoint asks for the application checks, and
one application check or more return an error, the endpoint returns
``503 ERROR`` and logs ``Health response: ERROR (<all-check-results>)``.


Application checks
==================

The application checks are routines in your application that subscribe to
``pyramid_health.HealthCheckEvent`` event, execute a specific health check and
report the outcome as a status (``OK`` or ``ERROR``) and an optional message.

The application checks are not called unless you explicitely request it with
the request param ``checks`` set to ``true`` or ``all`` (like:
``GET /health?checks=all``)

To add an application check in your application:

.. code-block:: python

   from pyramid.events import subscriber
   from pyramid_health import HealthCheckEvent


   @subscriber(HealthCheckEvent)
   def db_check(event):
       try:
           db.ping()
       except:
           event.report(name='db', status='NOK', message='ping failed')
       else:
           event.report(name='db', status='OK')

Notes:

- You may or may not report succeeding checks


Maintenance mode
================

In maintenance mode, the healthcheck endpoint's response is changed to inform
the HTTP client that this backend is unavailable. Typically a loadbalancer
polling the backends would stop sending traffic to a backend in maintenance
mode.

The response status code is ``299 MAINTENANCE`` by default. You can
change it with ``healthcheck.maintenance_code``.


CHANGES.rst
Changelog
=========

Development
-----------

1.2.1 (2015-06-06)
------------------

1.2.0 (2015-05-19)
------------------

* Change the default value for maintenance response code (from 503 to 299)

1.1.0 (2015-05-08)
------------------

* Add a query param to enable application checks (otherwise ignored)


1.0.0 (2015-05-08)
------------------

* Add application checks
* Add setting to change the health check url
* Add setting to change the maintenance response code
* Add setting to change the failure response code
* Add logging for maintenance and failure
* Update documentation
* Tests on Drone.io


0.3
---

* Change setting key for healthcheck.disablefile
* Enhance HTTP error message when healthcheck is disable by file


0.2
---

* Add option to enable maintenance mode when a file exists


0.1
---

* Basic health check view
* Public release


pyramid_health/tests.py
import unittest
import tempfile

import webtest
from pyramid.config import Configurator


class TestSimple(unittest.TestCase):

    def setup(self, url=None, maintenance_code=None, failure_code=None,
              disablefile=None):

        config = Configurator()

        if url is not None:
            config.add_settings({'healthcheck.url': url})
        if maintenance_code is not None:
            config.add_settings(
                {'healthcheck.maintenance_code': maintenance_code})
        if disablefile is not None:
            config.add_settings({'healthcheck.disablefile': disablefile})

        config.include('pyramid_health')

        self.app = webtest.TestApp(config.make_wsgi_app())

    def test_get(self):
        self.setup()
        response = self.app.get('/health', status=200)
        self.assertEqual(response.body, 'OK')

    def test_post(self):
        self.setup()
        response = self.app.post('/health', status=200)
        self.assertEqual(response.body, 'OK')

    def test_url_get(self):
        self.setup(url='/whatsup')
        response = self.app.get('/whatsup', status=200)
        self.assertEqual(response.body, 'OK')

    def test_get_maintenance_on(self):
        tmpfile = tempfile.NamedTemporaryFile()
        self.setup(disablefile=tmpfile.name)

        response = self.app.get('/health', status=299)
        self.assertEqual(response.body, 'MAINTENANCE')

    def test_get_maintenance_on_code(self):
        tmpfile = tempfile.NamedTemporaryFile()
        self.setup(disablefile=tmpfile.name, maintenance_code=288)

        response = self.app.get('/health', status=288)
        self.assertEqual(response.body, 'MAINTENANCE')

    def test_get_maintenance_off(self):
        tmpfile = tempfile.NamedTemporaryFile()
        self.setup(disablefile=tmpfile.name)

        tmpfile.close()  # Remove the disablefile

        response = self.app.get('/health', status=200)
        self.assertEqual(response.body, 'OK')


class TestChecks(unittest.TestCase):

    def setup(self, failure_code=None):
        from pyramid_health import HealthCheckEvent

        self.check1_status = 'OK'
        self.check1_message = 'report1'

        def check1(event):
            event.report(name='check1',
                         status=self.check1_status,
                         message=self.check1_message)

        self.check2_status = 'OK'
        self.check2_message = 'report2'

        def check2(event):
            event.report(name='check2',
                         status=self.check2_status,
                         message=self.check2_message)
        config = Configurator()
        if failure_code is not None:
            config.add_settings({'healthcheck.failure_code': failure_code})
        config.include('pyramid_health')
        config.add_subscriber(check1, HealthCheckEvent)
        config.add_subscriber(check2, HealthCheckEvent)

        self.app = webtest.TestApp(config.make_wsgi_app())

    def test_param(self):
        self.setup()
        self.check1_status = 'NOK'

        response = self.app.get('/health?checks=all', status=503)
        self.assertEqual(response.body, 'ERROR')

        response = self.app.get('/health?checks=true', status=503)
        self.assertEqual(response.body, 'ERROR')

        response = self.app.get('/health', status=200)
        self.assertEqual(response.body, 'OK')

    def test_ok(self):
        self.setup()
        response = self.app.get('/health?checks=all', status=200)
        self.assertEqual(response.body, 'OK')

    def test_check1_nok(self):
        self.setup()
        self.check1_status = 'NOK'
        self.check1_message = 'kaputt!'

        response = self.app.get('/health?checks=all', status=503)
        self.assertEqual(response.body, 'ERROR')

    def test_check2_nok(self):
        self.setup()
        self.check2_status = 'NOK'
        self.check2_message = 'nope'

        response = self.app.get('/health?checks=all', status=503)
        self.assertEqual(response.body, 'ERROR')

    def test_no_message(self):
        self.setup()
        self.check2_status = 'NOK'
        self.check2_message = None

        response = self.app.get('/health?checks=all', status=503)
        self.assertEqual(response.body, 'ERROR')

    def test_all_nok(self):
        self.setup()
        self.check1_status = 'NOK'
        self.check2_status = 'NOK'

        response = self.app.get('/health?checks=all', status=503)
        self.assertEqual(response.body, 'ERROR')

    def test_nok_failure_code(self):
        self.setup(failure_code=500)

        self.check1_status = 'NOK'

        response = self.app.get('/health?checks=all', status=500)
        self.assertEqual(response.body, 'ERROR')


pyramid_health/__init__.py
import os
import logging

log = logging.getLogger(__name__)


class HealthCheckEvent(object):
    """ Pyramid_health health check event """

    def __init__(self):
        self.checks = []

    def report(self, name, status, message=None):
        """ Report the result of an application check

        Args:
            name (str): Name of the application check
            status (str): Check status. Must be either "OK" or "ERROR"
            message (str): Optional detail about the check result
        """
        self.checks.append({
            'name': name,
            'status': status,
            'message': message
            })

    @property
    def status(self):
        """ Return 'OK' if all checks are OK and if no checks were reported """
        # all([]) is True
        if all([check['status'] == 'OK' for check in self.checks]):
            return 'OK'
        else:
            return 'ERROR'


def includeme(config):
    settings = config.registry.settings
    url = settings.get('healthcheck.url', '/health')
    disablefile = settings.get('healthcheck.disablefile', False)
    maintenance_code = int(settings.get('healthcheck.maintenance_code', 299))
    failure_code = int(settings.get('healthcheck.failure_code', 503))

    def health(request):
        if disablefile and os.path.exists(disablefile):
            log.info("Health response: MAINTENANCE")
            request.response.status_code = maintenance_code
            return 'MAINTENANCE'

        checks_param = request.params.get('checks')
        if checks_param and checks_param.lower() in ['true', 'all']:
            event = HealthCheckEvent()
            request.registry.notify(event)

            if event.status == 'ERROR':
                log.error("Health response: ERROR (%s)", event.checks)
                request.response.status_code = failure_code
                return 'ERROR'

        return 'OK'

    config.add_route('healthcheck', url)
    config.add_view(health, route_name='healthcheck', renderer='string')
